Add-SrmLicenseKey is failing with Could not find any of the servers specified by name

### Code of Conduct

- [X] I have read and agree to the [Code of Conduct](https://github.com/vmware/power-validated-solutions-for-cloud-foundation/blob/main/CODE_OF_CONDUCT.md).
- [X] Vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue initial description to help the maintainers prioritize.
- [X] Do not leave "+1" or other comments that do not add relevant information or questions.
- [X] If you are interested in working on this issue or have submitted a pull request, please leave a comment.

### VMware Cloud Foundation

5.1.1

### Module Version

2.10.0.1026

### PowerShell Version

7.4.2

### PowerCLI Version

13.2.1.22851661

### PowerVCF Version

2.4.1.1000

### Guest Operating System

Windows Server 2022

### Environment Details

Standard nested internal lab

### Description

When executing the `Add-SrmLicenseKey` cmdlet it sometimes errors

### Error or Debug Output

```powershell
[05-03-2024_12:12:31] [INFO] Assigning Licenses to Site Recovery Manager for Site Protection and Disaster Recovery for VMware Cloud Foundation
[05-03-2024_12:12:36] [INFO]  Error at Script Line 6940
[05-03-2024_12:12:36] [INFO]  Relevant Command: Disconnect-SrmServer -Server $srmFqdn -Force -Confirm:$false
[05-03-2024_12:12:36] [INFO]  Error Message: 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] Unable to connect to the remote server: dr.fault.SessionLimitExceeded
[05-03-2024_12:12:36] [ERROR] Unable to connect to the remote server: dr.fault.SessionLimitExceeded
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find SrmServer with name 'sfo-m01-srm01.sfo.rainpole.io'.	
[05-03-2024_12:12:36] [ERROR] System.Management.Automation.CmdletInvocationException: 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
 ---> VMware.VimAutomation.Sdk.Types.V1.ErrorHandling.VimException.ServerObnFailureException: 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
   at VMware.VimAutomation.Sdk.Util10Ps.BaseCmdlet.ObnSelector.TerminateObnProcessingDueToServerObnFailure(String errorId, ErrorCategory errorCategory, Object targetObject, Exception innerException, String connectionId, Object[] messageParams)
   at VMware.VimAutomation.Sdk.Util10Ps.BaseCmdlet.ObnSelector.SetParameterNewValue(PropertyInfo parameter, ArrayList expandedItemList)
   at VMware.VimAutomation.Sdk.Util10Ps.BaseCmdlet.ObnSelector.SelectObjectByName(PropertyInfo parameter)
   at VMware.VimAutomation.Sdk.Util10Ps.BaseCmdlet.ObnSelector.SelectServerByName()
   at VMware.VimAutomation.Sdk.Util10Ps.BaseCmdlet.BaseCmdlet.SelectServerByName()
   at VMware.VimAutomation.Sdk.Util10Ps.BaseCmdlet.BaseCmdlet.ProcessRecordErrorHandled()
   at VMware.VimAutomation.Srm.Commands.Commands.BaseSrmCmdlet.ProcessRecordErrorHandled()
   --- End of inner exception stack trace ---
   at System.Management.Automation.ExceptionHandlingOps.CheckActionPreference(FunctionContext funcContext, Exception exception)
   at System.Management.Automation.Interpreter.ActionCallInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
[05-03-2024_12:12:36] [ERROR] 5/3/2024 12:12:36	Disconnect-SrmServer		Could not find any of the servers specified by name.	
```

### Expected Behavior

Adding a license should complete without any issue or report a single POST_VALIDATION_FAILURE

### Actual Behavior

Failed

### Steps to Reproduce

Execute `Add-SrmLicenseKey` on the command line

### Log Fragments and Files

_No response_

### Screenshots

_No response_

### References

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -58,6 +58,8 @@
 - Fixed `Invoke-GlobalWsaDeployment` cmdlet to handle single and multiple nodes when using `Add-ClusterGroup` with Workspace ONE Access.
 - Fixed `Set-vCenterPermission` cmdlet to better handle expected errors.
 - Fixed `Remove-VrmsReplication` cmdlet where it was calling an incorrect name for `Get-VrmsReplication`.
+- Fixed `Add-SrmLicenseKey` cmdlet which was failing due to incorrect placement of `Disconnect-SrmServer` command.
+- Fixed `Undo-SrmLicenseKey` cmdlet which was not issuing a `Disconnect-SrmServer` command.
 - Enhanced `Add-NsxtIdentitySource` cmdlet to verify the Active Directory credentials are valid.
 - Enhanced `Invoke-UndoPcaDeployment` cmdlet to remove the VM folder for Private Cloud Automation.
 - Enhanced `Invoke-HrmDeployment` cmdlet to set the $failureDetected variable to false before starting the deployment.

--- a/PowerValidatedSolutions.psd1
+++ b/PowerValidatedSolutions.psd1
@@ -11,7 +11,7 @@
     RootModule = 'PowerValidatedSolutions.psm1'
 
     # Version number of this module.
-    ModuleVersion = '2.10.0.1027'
+    ModuleVersion = '2.10.0.1028'
     # Supported PSEditions
     # CompatiblePSEditions = @()
 

--- a/PowerValidatedSolutions.psm1
+++ b/PowerValidatedSolutions.psm1
@@ -6904,8 +6904,7 @@ Function Add-SrmLicenseKey {
                 if (($vcfVcenterDetails = Get-vCenterServerDetail -server $server -user $user -pass $pass -domain $domain)) {
                     if (Test-VsphereConnection -server $($vcfVcenterDetails.fqdn)) {
                         if (Test-VsphereAuthentication -server $vcfVcenterDetails.fqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
-                            if ((((Get-View -server $vcfVcenterDetails.fqdn ExtensionManager).ExtensionList | Where-Object { $_.key -eq "com.vmware.vcDr" }).Server.Url -Split "//" -Split ":")[2]) {
-                                $srmFqdn = (((Get-View -server $vcfVcenterDetails.fqdn ExtensionManager).ExtensionList | Where-Object { $_.key -eq "com.vmware.vcDr" }).Server.Url -Split "//" -Split ":")[2]
+                            if ($srmFqdn = Test-SrmRegistration -server $vcfVcenterDetails.fqdn) {
                                 if (Test-SrmConnection -server $srmFqdn) {
                                     if (Test-SrmAuthentication -server $srmFqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
                                         $serviceInstance = Get-View -Server $vcfVcenterDetails.fqdn ServiceInstance
@@ -6938,11 +6937,9 @@ Function Add-SrmLicenseKey {
                                                 }
                                             }
                                         }
+                                        Disconnect-SrmServer -Server $srmFqdn -Force -Confirm:$false
                                     }
-                                    Disconnect-SrmServer -Server $srmFqdn -Force -Confirm:$false
                                 }
-                            } else {
-                                Write-Error "No Site Recovery Manager Appliance Registered with vCenter Server ($($vcfVcenterDetails.fqdn))"
                             }
                         }
                         Disconnect-VIServer * -Force -Confirm:$false -WarningAction SilentlyContinue
@@ -7003,8 +7000,7 @@ Function Undo-SrmLicenseKey {
                 if (($vcfVcenterDetails = Get-vCenterServerDetail -server $server -user $user -pass $pass -domain $domain)) {
                     if (Test-VsphereConnection -server $($vcfVcenterDetails.fqdn)) {
                         if (Test-VsphereAuthentication -server $vcfVcenterDetails.fqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
-                            if ((((Get-View -server $vcfVcenterDetails.fqdn ExtensionManager).ExtensionList | Where-Object { $_.key -eq "com.vmware.vcDr" }).Server.Url -Split "//" -Split ":")[2]) {
-                                $srmFqdn = (((Get-View -server $vcfVcenterDetails.fqdn ExtensionManager).ExtensionList | Where-Object { $_.key -eq "com.vmware.vcDr" }).Server.Url -Split "//" -Split ":")[2]
+                            if ($srmFqdn = Test-SrmRegistration -server $vcfVcenterDetails.fqdn) {
                                 if (Test-SrmConnection -server $srmFqdn) {
                                     if (Test-SrmAuthentication -server $srmFqdn -user $vcfVcenterDetails.ssoAdmin -pass $vcfVcenterDetails.ssoAdminPass) {
                                         $serviceInstance = Get-View -Server $vcfVcenterDetails.fqdn ServiceInstance
@@ -7037,6 +7033,7 @@ Function Undo-SrmLicenseKey {
                                                 }
                                             }
                                         }
+                                        Disconnect-SrmServer -Server $srmFqdn -Force -Confirm:$false
                                     }
                                 }
                             }
@@ -53549,46 +53546,17 @@ Function Test-SrmAuthentication {
         [Parameter (Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$remotePass
     )
 
-    Try {
-        if ($remoteUser -and $remotePass) {
-            Connect-SrmServer -SrmServerAddress $server -user $user -password $pass -remoteUser $remoteUser -remotePassword $remotePass | Out-Null
-        } else {
-            Connect-SrmServer -SrmServerAddress $server -user $user -password $pass | Out-Null
-        }
-    } Catch {
-        if ($_.Exception.Message -eq "Cannot complete login due to an incorrect user name or password.") {
-            $PSCmdlet.ThrowTerminatingError(
-                [System.Management.Automation.ErrorRecord]::new(
-                    ([System.Security.Authentication.InvalidCredentialException]"Unable to authenticate with Site Recovery Manager ($server), check credentials: PRE_VALIDATION_FAILED"),
-                    'Test-SrmAuthentication',
-                    [System.Management.Automation.ErrorCategory]::AuthenticationError,
-                    ""
-                )
-            )
-        } elseif ($_.Exception.Message -eq "Unable to connect to the remote server: One or more errors occurred.") {
-            $PSCmdlet.ThrowTerminatingError(
-                [System.Management.Automation.ErrorRecord]::new(
-                    ([System.Management.Automation.ProviderNotFoundException]"Unable to authenticate with Site Recovery Manager ($server), check server IP address/FQDN: PRE_VALIDATION_FAILED"),
-                    'Test-SrmAuthentication',
-                    [System.Management.Automation.ErrorCategory]::InvalidArgument,
-                    ""
-                )
-            )
-        } elseif ($_.Exception.Message -eq "Unable to connect to the remote server: The connection to the remote server is down.") {
-            $PSCmdlet.ThrowTerminatingError(
-                [System.Management.Automation.ErrorRecord]::new(
-                    ([System.Management.Automation.RemoteException]"Unable to authenticate with the remote Site Recovery Manager instance: PRE_VALIDATION_FAILED"),
-                    'Test-SrmAuthentication',
-                    [System.Management.Automation.ErrorCategory]::ConnectionError,
-                    ""
-                )
-            )
-        }
+    if ($remoteUser -and $remotePass) {
+        Connect-SrmServer -SrmServerAddress $server -user $user -password $pass -remoteUser $remoteUser -remotePassword $remotePass | Out-Null
+    } else {
+        Connect-SrmServer -SrmServerAddress $server -user $user -password $pass | Out-Null
     }
+
     if ($defaultSrmServers) {
         $srmAuthentication = $True
         Return $srmAuthentication
     } else {
+        Write-Error "Unable to authenticate to Site Recovery Manager ($server), check credentials: PRE_VALIDATION_FAILED"
         $srmAuthentication = $False
         Return $srmAuthentication
     }

